### PR TITLE
soc: infineon: Support for power management on 20829

### PR DESCRIPTION
**Pull request by sreeramIfx**
- Initial changes in board, dts, and soc files to support system power management

### CODE REVIEW

**Review by ifyall — DISMISSED** [review]

File: boards/infineon/cyw920829m2evk_02/cyw920829m2evk_02.dts [hunk 1]
```diff
@@ -24,6 +24,26 @@
...
+			power-state-name = "suspend-to-ram";
+			min-residency-us = <2000000>;
+		};
+	};
+
+	cpus {
+		cpu@0 {
+			cpu-power-states = <&idle &suspend_to_ram>;
+		};
+	};
```
Comment (gmarull) [inline boards/infineon/cyw920829m2evk_02/cyw920829m2evk_02.dts:46]: power states need to be part of soc dts files

**Review by gmarull — CHANGES_REQUESTED** [review]

File: boards/infineon/cyw920829m2evk_02/cyw920829m2evk_02.dts (hunk 1 above)
Comment (sreeramIfx) [reply boards/infineon/cyw920829m2evk_02/cyw920829m2evk_02.dts]: moved the power states to soc dtsi files..

Comment (ifyall) [reply boards/infineon/cyw920829m2evk_02/cyw920829m2evk_02.dts]: @gmarull, please review Sreeram's change at your convenience.

**Comment by sreeramIfx** [comment]
Build failure is addressed @ https://github.com/zephyrproject-rtos/zephyr/pull/79399

File: soc/infineon/cat1b/cyw20829/power.c [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+	/* System Domain Idle Power Mode Configuration */
+	Cy_SysPm_SetDeepSleepMode(CY_SYSPM_MODE_DEEPSLEEP);
+
+#endif
+
+	return cyhal_syspm_init();
+}
+
+SYS_INIT(ifx_pm_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```
Comment (gmarull) [inline soc/infineon/cat1b/cyw20829/power.c:106]: please, use hooks

File: dts/arm/infineon/cat1b/cyw20829/cyw20829.dtsi [hunk 3]
```diff
@@ -16,6 +16,21 @@
 			device_type = "cpu";
 			compatible = "arm,cortex-m33";
 			reg = <0>;
+			cpu-power-states = <&idle &suspend_to_ram>;
+		};
+	};
+
+	power-states {
```
Comment (gmarull) [inline dts/arm/infineon/cat1b/cyw20829/cyw20829.dtsi:23]: not in cpus?

**Review by ifyall — DISMISSED** [review]

File: soc/infineon/cat1b/cyw20829/power.c (hunk 2 above)
Comment (sreeramIfx) [reply soc/infineon/cat1b/cyw20829/power.c]: updated to use hooks

File: dts/arm/infineon/cat1b/cyw20829/cyw20829.dtsi (hunk 3 above)
Comment (sreeramIfx) [reply dts/arm/infineon/cat1b/cyw20829/cyw20829.dtsi]: moved into the cpu node

File: soc/infineon/cat1b/cyw20829/CMakeLists.txt [hunk 4]
```diff
@@ -5,6 +5,8 @@ zephyr_sources(soc.c)
 zephyr_sources(app_header.c)
 zephyr_include_directories(.)
 
+zephyr_sources_ifdef(CONFIG_PM ./power.c)
```
Comment (gmarull) [inline soc/infineon/cat1b/cyw20829/CMakeLists.txt:8]:
```suggestion
zephyr_sources_ifdef(CONFIG_PM power.c)
```

File: dts/arm/infineon/cat1b/cyw20829/cyw20829.dtsi [hunk 5]
```diff
@@ -16,9 +16,25 @@
...
+
+			suspend_to_ram: suspend_to_ram {
+				compatible = "zephyr,power-state";
+				power-state-name = "suspend-to-ram";
+				min-residency-us = <2000000>;
+			};
 		};
 	};
 
+
```
Comment (gmarull) [inline dts/arm/infineon/cat1b/cyw20829/cyw20829.dtsi:37]:
```suggestion
```

File: soc/infineon/cat1b/cyw20829/Kconfig.defconfig [hunk 6]
```diff
@@ -6,6 +6,13 @@
 
 if SOC_DIE_CYW20829
 
+config INFINEON_CAT1_LP_TIMER
+	bool
+	default n
```
Comment (gmarull) [inline soc/infineon/cat1b/cyw20829/Kconfig.defconfig:11]:
```suggestion
```

File: soc/infineon/cat1b/cyw20829/power.c [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+	*deepsleepcounts = deepsleep_counts;
+	*sleeptime = last_sleep_duration;
+}
+
+/*
+ * Called from pm_system_suspend(int32_t ticks) in subsys/power.c
+ * For deep sleep pm_system_suspend has executed all the driver
+ * power management call backs.
+ */
+__weak void pm_state_set(enum pm_state state, uint8_t substate_id)
```
Comment (gmarull) [inline soc/infineon/cat1b/cyw20829/power.c:40]:
```suggestion
void pm_state_set(enum pm_state state, uint8_t substate_id)
```

File: soc/infineon/cat1b/cyw20829/power.c [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+		 */
+		SCB_SCR &= (uint32_t)~SCB_SCR_SLEEPDEEP_Msk;
+		break;
+	default:
+		LOG_DBG("Unsupported power state %u", state);
+		break;
+	}
+}
+
+__weak void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
```
Comment (gmarull) [inline soc/infineon/cat1b/cyw20829/power.c:79]:
```suggestion
void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
```

File: soc/infineon/cat1b/cyw20829/power.c [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+		break;
+
+	default:
+		break;
+	}
+}
+
+int ifx_pm_init(void)
+{
+#if defined(CONFIG_BOARD_CYW920829M2EVK_02)
```
Comment (gmarull) [inline soc/infineon/cat1b/cyw20829/power.c:96]: -1 board selective code in soc, find a better way

File: soc/infineon/cat1b/cyw20829/soc.c [hunk 10]
```diff
@@ -94,3 +97,10 @@ static int init_cycfg_platform_wrapper(void)
 }
 
 SYS_INIT(init_cycfg_platform_wrapper, PRE_KERNEL_1, 0);
+
+void soc_early_init_hook(void)
+{
+#ifdef CONFIG_PM
+	ifx_pm_init();
+#endif
```
Comment (gmarull) [inline soc/infineon/cat1b/cyw20829/soc.c:105]: ifdefout the entire hook?

**Review by gmarull — DISMISSED** [review]

**Review by ifyall — DISMISSED** [review]

File: soc/infineon/cat1b/cyw20829/power.h [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __IFX_POWER_H
+#define __IFX_POWER_H
+
+void get_sleep_info(uint32_t *sleepattempts, uint32_t *sleepcounts, uint32_t *deepsleepattempts,
+		    uint32_t *deepsleepcounts, int64_t *sleeptime);
+
```
[the remark this answers is not in the document]
Comment (gmarull) [reply soc/infineon/cat1b/cyw20829/power.h]: ping, impl still there

[the remark this answers is not in the document]
Comment (sreeramIfx) [reply soc/infineon/cat1b/cyw20829/power.h]: removed the implementation as well, please review

**Review by ifyall — APPROVED** [review]

**Review by npal-cy — APPROVED** [review]

**Pull request merged**